### PR TITLE
Opt-In Flodgatt in docker-compose.yml

### PR DESCRIPTION
**Pull request by shleeable**
This is my configuration currently running on aus.social using the precompiled binary.
If we want to compile from source instead, I'll make a patch to build using the Rust compiler later.

* Hey @Gargron, this is Opt-In. I'm happy for this to be merged soon after some additional testing and approval from yourself/other stakeholders, and you can default it later.
* Everybody else, Please feel free to test this on your environment and keep @codesections updated with any errors or issues. https://github.com/tootsuite/flodgatt

**Upgrade**: 
* Comment the NodeJS streaming part from docker-compose.yml
* Uncomment the Flodgatt streaming part from docker-compose.yml
* `docker-compose build flodgatt --pull`
* `docker-compose up -d --remove-orphans`

**Rollback**: 
* Uncomment the NodeJS streaming part from docker-compose.yml
* Comment the Flodgatt streaming part from docker-compose.yml
* `docker-compose up -d --remove-orphans`

### CODE REVIEW

**Comment by shleeable** [comment]
image
Zero complaints over the last 12+ hours.

File: Dockerfile.flodgatt [hunk 1]
```diff
@@ -0,0 +1,37 @@
+FROM ubuntu:18.04
+
+# Create the mastodon user
+ARG UID=991
+ARG GID=991
+RUN apt update && \
+	echo "Etc/UTC" > /etc/localtime && \
+	apt install -y whois wget && \
```
Comment (SuperSandro2000) [inline Dockerfile.flodgatt]:
```suggestion
	apt install --no-install-recommends -y whois wget && \
```

File: Dockerfile.flodgatt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	addgroup --gid $GID mastodon && \
+	useradd -m -u $UID -g $GID -d /opt/mastodon mastodon && \
+	echo "mastodon:`head /dev/urandom | tr -dc A-Za-z0-9 | head -c 24 | mkpasswd -s -m sha-256`" | chpasswd && \
```
Comment (SuperSandro2000) [inline Dockerfile.flodgatt]: Wouldn't it be easier to just lock the account?

File: Dockerfile.flodgatt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	rm -rf /var/cache && \
+	rm -rf /var/lib/apt/lists/*
+
+# Add tini
+ENV TINI_VERSION="v0.18.0"
+ENV TINI_SUM="12d20136605531b09a2c2dac02ccee85e1b874eb322ef6baf7561cd93f93c855"
+ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /tini
+RUN echo "$TINI_SUM tini" | sha256sum -c -
+RUN chmod +x /tini
```
Comment (SuperSandro2000) [inline Dockerfile.flodgatt]:
```suggestion
RUN echo "$TINI_SUM tini" | sha256sum -c - \
  && chmod +x /tini
```

File: Dockerfile.flodgatt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+# Install Flodgatt
+ENV FG_VER="v0.6.3"
+ENV FG_SUM="71eeae1784542edd06cd8b2aa35b3bbc87b037f197e866dfe67e451f99776622"
+ADD --chown=mastodon:mastodon https://github.com/tootsuite/flodgatt/releases/download/$FG_VER/flodgatt /flodgatt 
+RUN echo "$FG_SUM /flodgatt" | sha256sum -c -
+RUN chmod +x /flodgatt
```
Comment (SuperSandro2000) [inline Dockerfile.flodgatt]:
```suggestion
RUN echo "$FG_SUM /flodgatt" | sha256sum -c - \
  && chmod +x /flodgatt
```

**Review by SuperSandro2000 — COMMENTED** [review]
:)

**Comment by SuperSandro2000** [comment]
BTW Do we have any metrics on how much better floodgate performs?

File: Dockerfile.flodgatt (hunk 1 above)
Comment (shleeable) [reply Dockerfile.flodgatt]: I've taken this advice, and made some changes.

**Comment by codesections** [comment]
> BTW Do we have any metrics on how much better floodgate performs?

I don't have clear metrics for how it performs under load – that's one of the reasons I'd like to see Flodgatt tested more broadly.  The main area I'm focused on is reducing memory use (which is really the only resource the current streaming server much of).  On my test instance (with 4 test users, 0 real users, and a populated federated timeline) it uses ~2.5 MiB; for comparison, the Node streaming server uses ~35 MiB with similarly low load.

I doubt we'll see anywhere near that 10× reduction under load – much of the savings with zero load come from not having Node's runtime.  That said, I expect we will see a noticeable drop.  @shleeable posted the stats below for aus.social, presumably under typical load.  If so, that's well under Node's memory use for an empty server, which is pretty encouraging.  @shleeable, do you know how much memory the Node server would use under similar load?

flodgatt_ram_usage

**Comment by shleeable** [comment]
I've taken @SuperSandro2000s advice for some of the Dockerfile -- and ended up rewriting some of it.

File: Dockerfile.flodgatt (hunk 2 above)
Comment (shleeable) [reply Dockerfile.flodgatt]: Feel free to test and PR after this is merged... I'm leaving this for now

**Comment by codesections** [comment]
> I can drop back to nodeJS after that and record my baseline if you really want it.

No need; I was just curious.  I'm sure we'll have other before and after comparisons after a while.

**Comment by shleeable** [comment]
Tested above commits.. other people still need to test the whole thing tho

**Comment by SuperSandro2000** [comment]
I plan to test the Dockerfile in the following days but I can't test if the docker-compose file is right or not.
Right now my streaming docker container uses about 110MB with 21 days uptime. I will report back when I switched.

**Comment by shleeable** [comment]
image
After 3days

**Comment by shleeable** [comment]
Sorry for the mess. I've split the flodgatt and dockerfile changes up for clarity.

**Pull request closed without merging**